src/Symfony/Component/Security/Csrf/Tests/DoubleSubmitCsrfTokenManagerTest.php

```diff
@@ -0,0 +1,186 @@
+    {
+        $request = new Request();
+        $request->headers->set('Origin', $request->getSchemeAndHttpHost());
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', str_repeat('a', 24));
+
+        $this->logger->expects($this->once())->method('debug')->with('CSRF validation accepted using origin info.');
+        $this->assertTrue($this->csrfTokenManager->isTokenValid($token));
+        $this->assertSame('origin', $request->attributes->get('x-csrf-token'));
+    }
+
+    public function testValidOriginAfterDoubleSubmit()
+    {
+        $session = $this->createMock(Session::class);
+        $request = new Request();
+        $request->setSession($session);
+        $request->headers->set('Origin', $request->getSchemeAndHttpHost());
+        $request->cookies->set('sess', 'id');
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', str_repeat('a', 24));
+
+        $session->expects($this->once())->method('getName')->willReturn('sess');
+        $session->expects($this->once())->method('get')->with('x-csrf-token')->willReturn('double-submit');
+        $this->logger->expects($this->once())->method('warning')->with('CSRF validation failed: double-submit info was used in a previous request but is now missing.');
+        $this->assertFalse($this->csrfTokenManager->isTokenValid($token));
+    }
+
+    public function testValidDoubleSubmit()
+    {
+        $request = new Request();
+        $request->cookies->set('x-csrf-token_'.str_repeat('a', 24), 'x-csrf-token');
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', str_repeat('a', 24));
+
+        $this->logger->expects($this->once())->method('debug')->with('CSRF validation accepted using double-submit info.');
+        $this->assertTrue($this->csrfTokenManager->isTokenValid($token));
+        $this->assertSame('double-submit', $request->attributes->get('x-csrf-token'));
+    }
+
+    public function testCheckOnlyHeader()
+    {
+        $csrfTokenManager = new DoubleSubmitCsrfTokenManager($this->requestStack, $this->logger, DoubleSubmitCsrfTokenManager::CHECK_ONLY_HEADER);
+
+        $request = new Request();
+        $tokenValue = str_repeat('a', 24);
+        $request->headers->set('x-csrf-token', $tokenValue);
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', $tokenValue);
+
+        $this->logger->expects($this->once())->method('debug')->with('CSRF validation accepted using header info.');
+        $this->assertTrue($csrfTokenManager->isTokenValid($token));
+        $this->assertSame('x-csrf-token', $request->cookies->get('x-csrf-token_'.$tokenValue));
+
+        $this->logger->expects($this->once())->method('warning')->with('CSRF validation failed: wrong token found in header info.');
+        $this->assertFalse($csrfTokenManager->isTokenValid(new CsrfToken('test_token', str_repeat('b', 24))));
```

Comment:
Thanks for all the review notes.
Would you be up to submit more test cases [on my fork](https://github.com/nicolas-grekas/symfony/tree/csrf-double-submit)? I'd happily accept them :)

Reply:
It's the one that's implemented ;)

Reply:
Extracting the behaviour from the implementation is a very hard task to do.

Even with code coverage and mutation testing tools.

There will be exploratory tests then, even with them it will keep some holes on the actual behaviour.

Reply:
I added more test cases.

Reply:
Thank you @nicolas-grekas to had spent time to cover more behaviours.